### PR TITLE
Feature/sds support

### PR DESCRIPTION
**Pull request by pxzero**
### Issues associated with this PR

For #732 

### Solutions
Follow Istio citadel & citadel agent solution

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases.
And you need to show the ut result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/xds/sdsclient.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+var sdsClient *SdsClientImpl
+var sdsClientLock sync.Mutex
+
+// GetSdsClientImpl use by tls module , when get sds config from xds
+func GetSdsClient(config *auth.SdsSecretConfig) v2.SdsClient {
+	if sdsClient != nil {
+		return sdsClient
+	} else {
+		sdsClientLock.Lock()
+		defer sdsClientLock.Unlock()
```
Comment (nejisama) [inline pkg/xds/sdsclient.go]: 获取的时候就要拿锁，否则可能会重复创建

File: pkg/xds/sdsclient.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+			return nil
+		}
+		return sdsClient
+	}
+	return nil
+}
+
+// CloseSdsClientImpl used only mosn exit
+func CloseSdsClient() {
+	if sdsClient != nil && sdsClient.sdsSubscriber != nil {
```
Comment (nejisama) [inline pkg/xds/sdsclient.go]: 这里也需要锁

Comment (nejisama) [reply pkg/xds/sdsclient.go]: 同时Stop以后 感觉应该要把sdsClient设置为空

File: pkg/xds/sdsclient.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+		sdsClient.sdsSubscriber.Stop()
+	}
+}
+
+func (client *SdsClientImpl) AddUpdateCallback(sdsConfig *auth.SdsSecretConfig, callback v2.SdsUpdateCallbackFunc) {
+	client.updatedLock.Lock()
```
Comment (nejisama) [inline pkg/xds/sdsclient.go]: 需要判断client是不是空，下同

File: pkg/xds/sdsclient.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	defer client.updatedLock.Unlock()
+	delete(client.SdsConfigMap, sdsConfig.Name)
+	delete(client.SdsCallbackMap, sdsConfig.Name)
+	return nil
+}
+
+// SetSecret invoked when sds subscriber get secret response
+func (client *SdsClientImpl) SetSecret(name string, secret *auth.Secret) {
+	if fc, ok := client.SdsCallbackMap[name]; ok {
```
Comment (nejisama) [inline pkg/xds/sdsclient.go]: 锁

File: pkg/xds/sdsclient_test.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	}()
+	sdsClient.AddUpdateCallback(config, func(name string, secret *types.SDSSecret) {
+		if name != "default" {
+			t.Errorf("name should same with config.name")
+		}
+		updatedChan <- 1
+	})
+	select {
+	case <-updatedChan:
+	case <-timeoutChan:
```
Comment (nejisama) [inline pkg/xds/sdsclient_test.go]: 这里感觉可以直接time.After

File: pkg/xds/v2/sdssubscriber.go [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+func (subscribe *SdsSubscriber) Start() error {
+	sdsStreamConfig := subscribe.convertSdsConfig(subscribe.sdsConfig)
+	streamClient, err := subscribe.getSdsStreamClient(sdsStreamConfig)
+	if err != nil {
+		return err
+	}
+	subscribe.sdsStreamClient = streamClient
+	go subscribe.sendRequestLoop(subscribe.sdsStreamClient)
```
Comment (nejisama) [inline pkg/xds/v2/sdssubscriber.go]: 开启goroutine都存在panic风险，建议utils.GoWithRecover，目前mosn基本都采取这个方式

File: pkg/xds/v2/sdssubscriber.go [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+	go subscribe.receiveResponseLoop(subscribe.sdsStreamClient)
+	return nil
+}
+
+func (subscribe *SdsSubscriber) Stop() {
+	subscribe.sendStopChannel <- 0
+	subscribe.receiveStopChannel <- 0
+}
```
Comment (nejisama) [inline pkg/xds/v2/sdssubscriber.go]: 这里感觉可以用一个stop chan，通过close的方式来通知
close(StopChan)

File: pkg/xds/v2/sdssubscriber.go [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+				log.DefaultLogger.Errorf("[xds] [sds subscriber] only support one grpc service,but get %v", len(grpcService))
+				return nil
+			}
+			if grpcConfig, ok := grpcService[0].TargetSpecifier.(*core.GrpcService_GoogleGrpc_); ok {
+				sdsStreamConfig.sdsUdsPath = grpcConfig.GoogleGrpc.TargetUri
+				sdsStreamConfig.statPrefix = grpcConfig.GoogleGrpc.StatPrefix
+			}
+		}
+	}
+	return sdsStreamConfig
```
Comment (nejisama) [inline pkg/xds/v2/sdssubscriber.go]: 如果不是*core.ConfigSource_ApiConfigSource 也预期返回成功？而不是nil？和grpcService != 1 有什么区别？

File: pkg/xds/v2/sdssubscriber.go (hunk 6 above)
Comment (nejisama) [reply pkg/xds/v2/sdssubscriber.go]: 用utils.GoWithRecover 统一panic的输出，也没有必要在具体的函数里去做recover的输出

File: pkg/xds/v2/sdssubscriber.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+			return
+		default:
+			if subscribe.sdsStreamClient == nil {
+				log.DefaultLogger.Warnf("[xds] [sds subscriber] stream client closed, sleep 1s and wait for reconnect")
+				time.Sleep(time.Second)
+				continue
+			}
+			resp, err := subscribe.sdsStreamClient.streamSecretsClient.Recv()
+			if err != nil {
+				log.DefaultLogger.Warnf("[xds] [sds subscriber] get resp timeout: %v, retry after 1s", err)
```
Comment (nejisama) [inline pkg/xds/v2/sdssubscriber.go]: 这里感觉超时是不是常态？输出Warn日志会不会造成日志污染？

File: pkg/xds/v2/sdssubscriber.go [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+		}
+	}
+}
+
+func (subscribe *SdsSubscriber) sendRequest(request *xdsapi.DiscoveryRequest) error {
+	err := subscribe.sdsStreamClient.streamSecretsClient.Send(request)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (nejisama) [inline pkg/xds/v2/sdssubscriber.go]: 像这种输出，直接return Send()就可以了

File: pkg/xds/xdsclient.go [hunk 11]
```diff
@@ -45,6 +45,9 @@ type Client struct {
 	adsClient *v2.ADSClient
 }
 
+var ServiceNode = ""
+var ServiceCluster = ""
+
```
Comment (nejisama) [inline pkg/xds/xdsclient.go]: 这个全局变量是设置来做什么的？

File: pkg/xds/sdsclient.go (hunk 1 above)
Comment (pxzero) [reply pkg/xds/sdsclient.go]: fixed.thx

File: pkg/xds/sdsclient.go (hunk 2 above)
Comment (pxzero) [reply pkg/xds/sdsclient.go]: fixed.thx

File: pkg/xds/sdsclient.go (hunk 3 above)
Comment (pxzero) [reply pkg/xds/sdsclient.go]: fixed.thx

File: pkg/xds/sdsclient.go (hunk 4 above)
Comment (pxzero) [reply pkg/xds/sdsclient.go]: fixed.thx

File: pkg/xds/sdsclient_test.go (hunk 5 above)
Comment (pxzero) [reply pkg/xds/sdsclient_test.go]: fixed.thx

File: pkg/xds/v2/sdssubscriber.go (hunk 6 above)
Comment (pxzero) [reply pkg/xds/v2/sdssubscriber.go]: fixed.thx

File: pkg/xds/v2/sdssubscriber.go (hunk 7 above)
Comment (pxzero) [reply pkg/xds/v2/sdssubscriber.go]: fixed.thx

File: pkg/xds/v2/sdssubscriber.go (hunk 8 above)
Comment (pxzero) [reply pkg/xds/v2/sdssubscriber.go]: fixed.thx

File: pkg/xds/v2/sdssubscriber.go (hunk 9 above)
Comment (pxzero) [reply pkg/xds/v2/sdssubscriber.go]: stream secret 接口是长连接状态grpc 接口，server 端会确保没有response时，维持状态，正常使用场景下不会常态打印这个日志

File: pkg/xds/v2/sdssubscriber.go (hunk 10 above)
Comment (pxzero) [reply pkg/xds/v2/sdssubscriber.go]: fixed.thx

File: pkg/xds/xdsclient.go (hunk 11 above)
Comment (pxzero) [reply pkg/xds/xdsclient.go]: ServiceNode & ServiceCluster 是Istio 用于识别sidecar标识，后续其他模块也会需要引用该数据，因此设计为全局变量
